clash-prelude/src/Clash/XException/MaybeX.hs

```diff
@@ -0,0 +1,149 @@
+{-|
+Copyright  :  (C) 2022, QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Helpers to make 'Clash.XException.XException' explicit in the type system.
+Using these helpers can help programmers account for 'Clash.XException.XException's
+properly in blackbox models or tests. Note that none of these operations can be
+translated to HDL.
+
+-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Clash.XException.MaybeX
+  ( MaybeX(IsX, IsDefined)
+
+    -- * Construction
+  , toMaybeX
+  , hasXToMaybeX
+
+    -- * Deconstruction
+  , unsafeToA
+
+    -- * Operations
+  , andX
+  , orX
+  , maybeX
+  ) where
+
+import Prelude
+
+import Control.Applicative
+import Control.DeepSeq (NFData)
+import Control.Exception (throw)
+
+import Clash.XException (XException(..), isX, hasX)
+
+-- | Structure helping programmers to deal with 'Clash.XException.XException'
+-- values. For safety reasons it can't be constructed directly, but should be
+-- constructed using either 'pure' or 'toMaybeX'. After construction, it can be
+-- deconstructed using either 'IsX' or 'IsDefined'.
+data MaybeX a
+  = IsX_ String
+  -- ^ Upon construction, @a@ evaluated to 'Clash.XException.XException'
+  | IsDefined_ !a
+  -- ^ Upon construction, @a@ evaluated to a non-bottom WHNF
+
+instance Show a => Show (MaybeX a) where
+  showsPrec d = \case
+    IsX_ msg     -> showParen (d > 10) $ showString "IsX "       . showsPrec 11 msg
+    IsDefined_ a -> showParen (d > 10) $ showString "IsDefined " . showsPrec 11 a
+
+-- | Upon construction, @a@ evaluated to 'Clash.XException.XException'
+pattern IsX :: forall a. String -> MaybeX a
+pattern IsX msg <- IsX_ msg
+
+-- | Upon construction, @a@ evaluated to a non-bottom WHNF
+pattern IsDefined :: forall a. a -> MaybeX a
+pattern IsDefined a <- IsDefined_ a
+{-# COMPLETE IsX, IsDefined #-}
+
+instance Functor MaybeX where
+  fmap _f (IsX_ msg) = IsX_ msg
+  fmap f  (IsDefined_ a) = either IsX_ IsDefined_ (isX (f a))
```

Comment:
First of all, as discussed, we should document the `IsX` propagation of `Functor` and `Applicative`.

```suggestion
  fmap f  (IsDefined_ a) = pure $ isX $ f a
```
Exact same thing, but reusing the `pure` implementation.